File: src/p1.rs
#[test]
fn test() {
    let v1: Vec<i32> = vec![2, 7, 11, 15];
    assert_eq!(two_sum(v1, 9), vec![0, 1]);

    let v2: Vec<i32> = vec![3, 2, 4];
    assert_eq!(two_sum(v2, 6), vec![1, 2]);
}

pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    use std::collections::HashMap;
    let mut map: HashMap<i32, i32> = HashMap::new();

    for (i, num) in nums.into_iter().enumerate() {
        match map.get(&(target - num)) {
            Some(&x) => return vec![x, i as i32],
            None => map.insert(num, i as i32),
        };
    }

    vec![]
}

File: src/p1.py
import unittest
from typing import Dict, List


# 一定存在下标为i, j的两个数字满足要求，枚举j，去哈希表里找target-nums[j]
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        map: Dict[int, int] = {}  # 数字->下标

        for j, num in enumerate(nums):
            if target - num in map:
                return [map[target - num], j]  # return [i, j]
            map[num] = j

        return []


class Test(unittest.TestCase):
    def test(self) -> None:
        solution = Solution()
        self.assertEqual(solution.twoSum([2, 7, 11, 15], 9), [0, 1])
        self.assertEqual(solution.twoSum([3, 2, 4], 6), [1, 2])
        self.assertEqual(solution.twoSum([3, 3], 6), [0, 1])


if __name__ == "__main__":
    unittest.main()
